### DISCUSSION TITLE
Can't manage to get file upload working in keystone

### DISCUSSION THREAD

**Discussion by coolcomfort** [keystonejs/keystone#7362] [Questions] [answered]
I am trying to get file uploading working in a typeDef. I am using extendGraphqlSchema, which allows me to populate my typeDef with the following code:

```
export const extendGraphqlSchema = graphQLSchemaExtension({
  typeDefs: graphql`
    type Mutation {
      uploadPost(title: String!, photo: Upload ): Post
    }
  `,
  resolvers: {
    Mutation: {
      uploadPost,
    },
  },
});
```

Notice how it is optional to upload a photo. Now, when I try to bring this inside 

```
async function uploadPost(
  root: any,
  { title, photo }: { title: string; photo: any},
  context: KeystoneContext
): Promise<PostCreateInput> {

  const sesh = context.session as Session;

  if (!sesh.itemId) {
    throw new Error("You must be logged in to do this!");
  }

  console.log("base area");

  if (photo) {

    console.log("arriving at photo upload area");

    createPost = await context.lists.Post.createOne({
      data: {
        title,
        photo: { upload: photo }
      },
      resolveFields: "id title",
    });
    
  } else {

    console.log("arriving at normal post area");

    createPost = await context.lists.Post.createOne({
      data: {
        title,
      },
      resolveFields: "id title",
    });
  }

  return createPost;
}
```

When I try to upload a file on the frontend, I get this: 
```Error: Variable "$item" got invalid value {} at "item.photo.upload"; Upload value invalid.```

I have tried changing `photo: { upload: photo }` to just `photo: photo` (aka photo), but nothing works. 

Here is what's going on in my frontend:

The graphQL code: 

```
const CREATE_POST = gql`
  mutation CREATE_POST(
    # which variables get passed in and what types
    $title: String!
    $photo: Upload
  ) {
    createPost(data: { title: $title, photo: { upload: $photo } }) {
      id
    }
  }
`;
```

The destructured statement: 

```
const [createPost, { loading, error, data }] = useMutation(CREATE_POST, {
    variables: {
      title: inputs.title,
      photo: inputs.photo
    },
    refetchQueries: [{ query: ALL_POSTS_QUERY }],
  });
```

I understand with the way the code is now, you might not think I need a typeDef for this. However, I will be adding more things to the database and I would like it to happen in one swift go rather than trying to make it work in the front end. 

EDIT: Just wanted to add, yes, I do have an adapter statement in my index:
```    images: {
      upload: 'local',
      local: {
        storagePath: '../frontend/public/images',
        baseUrl: '/images',
      },
    },
```

I created a test page for uploading photos without typeDefs just to ensure it works.

**Comment by coolcomfort** [comment]
The alternative to this is to create a schema solely for uploaded files, then do all file uploads on the front end, then pass that information to the typedef for any backend related purposes.

**Comment by emmatown** [comment, marked as answer]
You need to create a new `Upload` object to provide to `context.query` like this:

```ts
import { Upload } from 'graphql-upload';


// in your resolver
const upload = new Upload();
photo
  .catch(reason => {
    upload.reject(reason);
  })
  .then(data => {
    if (data !== undefined) {
      upload.resolve(data);
    }
  });
context.query.Post.createOne({
  data: {
    photo: { upload },
  },
});
```
